Explain how to switch to add native namespace

gforcada opened the pull request:
To go along with the https://github.com/plone/Products.CMFPlone/issues/3928

Disclaimer: I have no idea where would be the best place to put this documentation within the structure, and probably the admonitions style is not the right one.

We can polish, but before nitpicking on this details, let's agree on the structure of the document and if it explains all the relevant steps and are clear enough.

Once that's done we can do the polish.

@mauritsvanrees @davisagli: given a random Plone contributor, do you think the steps cover all the steps, and are clear enough? I made it generic enough (with tips to skip sections) so that it can be reused for `collective` or private packages, where we can not convert all ourselves.


<!-- readthedocs-preview plone6 start -->
----
📚 Documentation preview 📚: https://plone6--1981.org.readthedocs.build/

<!-- readthedocs-preview plone6 end -->

gforcada committed:
feat: explain how to switch to add native namespace

To go along with the PLIP 3928.

stevepiercy commented:
@gforcada this is a how-to guide for developers, so it belongs in `docs/developer-guide` with a toctree entry in `docs/developer-guide/index`. Would you please move it there?

I'll do a thorough non-technical review after the technical structure and details are reviewed and approved.

Thank you!

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
```diff
@@ -0,0 +1,171 @@
+# create a branch for it
+git checkout -b $MAJOR.x
+# push the newly created branch
+git push $MAJOR.x
```
This complains that for example 4.x is not a remote.  This seems enough:

```suggestion
git push
```

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+
+```shell
+# list all tags
+git for-each-ref --sort=taggerdate --format '%(tag)'
````
I have not used the `for-each-ref` command before, thanks for sharing.
When I try it, I get some empty lines as well, for references in `refs/heads` and `refs/remotes`.  We can restrict to only tags:

```suggestion
git for-each-ref --sort=taggerdate --format '%(tag) refs/tags
```

gforcada replied:
Oh nice, I did not know that you can further filter by refs 🎉

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
```diff
@@ -0,0 +1,171 @@
+# list all tags
+git for-each-ref --sort=taggerdate --format '%(tag)'
+# get the last tag's major number
+MAJOR=`git for-each-ref --sort=taggerdate --format '%(tag)' | tail -n1 | cut -d"." -f1`
```
Same here for good measure, only tags:

```suggestion
MAJOR=`git for-each-ref --sort=taggerdate --format '%(tag)' refs/tags | tail -n1 | cut -d"." -f1`
```

mauritsvanrees commented on docs/developer-guide/native-namespaces.md:
```diff
@@ -0,0 +1,171 @@
+!!! note
+    Only relevant for Plone core packages
+
+Clone the repository or ensure you are at latest changes:
```
I would add a line of explanation:

```suggestion
This part is only needed when the `main` or `master` branch is used on multiple versions of the Plone core development buildout.

Clone the repository or ensure you are at latest changes:
```

gforcada replied:
Well, I add the `Only relevant for Plone core packages, but sure, extra tips can never hurt :)

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+git push
+```
+
+!!! note
````
```suggestion
Now check if you need to do the same on the 6.0 branch of `buildout.coredev`.

!!! note
```

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
```diff
@@ -0,0 +1,171 @@
+
+### Numbers before
+
+One risk of changing to the native namespaces is that some files, or tests, might not be left behind.
```
I suppose the main risk of this was actually already when moving to a src-layout.

gforcada replied:
oh, right, so we can drop this parts then, which makes it all much simpler and straightforward 😄

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+To get the number of files on them run the following commands:
+
+```shell
+python -c "import tarfile; print(len(tarfile.open('dist/$package.tar.gz', 'r:gz').getnames()))"
````
```suggestion
python -c "import glob; import tarfile; print(len(tarfile.open(glob.glob('dist/*.tar.gz')[0], 'r:gz').getnames()))"
```

gforcada replied:
Should we keep this section or remove it, as your previous suggestion was that it was not that problematic anymore, as we already moved to the `src/` layout?

🤔 as the idea is to make it relevant for collective and private repositories, we might keep it, so it also helps there, where they might not have moved to a `src` layout.

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+
+```shell
+python -c "import tarfile; print(len(tarfile.open('dist/$package.tar.gz', 'r:gz').getnames()))"
+python -c "from zipfile import ZipFile; print(len(ZipFile('dist/$package.whl').namelist()))"
````
```suggestion
python -c "import glob; import zipfile; print(len(zipfile.ZipFile(glob.glob('dist/*.whl')[0]).namelist()))"
```

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+python -c "from zipfile import ZipFile; print(len(ZipFile('dist/$package.whl').namelist()))"
+```
+
+If all numbers match, review the changes once more, push the branch with the changes for others to review it.
````
There will be a few less files in the new distributions.

```suggestion
It is okay if the numbers are slightly lower.
For obvious reasons the old source distribution will have one or more extra `__init__.py` files.
Both old distributions are expected to have an extra `namespace_packages.txt` file and possible an `x-nspkg.pth` file.
If you lose more than a few files though, something is wrong.
If the numbers are close enough, review the changes once more, and push the branch with the changes for others to review it.
```

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+And update the scaffolding files with `plone.meta`:
+
+```shell
+uvx --from plone.meta config-package .
````
In my initial comment in https://github.com/plone/meta/pull/297 I first run config-package passing a specific branch name matching the one used by the pep-420 script, then update the test matrix, run config-package again, and finally run the pep-420 script.

You run the pep-420 script first, which I think makes sense.  But then we should tell config-package to use the current branch:


```suggestion
uvx --from plone.meta config-package --branch current .
```

mauritsvanrees commented on docs/reference-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+Get the list of tests, like before and compare the lists to ensure that the same amount of tests are found.
+
+```shell
+tox run -e test -- --list-tests
````
I think we can just get the number of lines:

```suggestion
tox run -e test -- --list-tests | wc -l
```

mauritsvanrees commented on docs/developer-guide/native-namespaces.md:
````diff
@@ -0,0 +1,171 @@
+cd $package
+uvx --from plone.meta switch-to-pep420 .
+```
+
````
```suggestion

Note that this will also bump the version to a new major release.
If the `main` or `master` branch is already an alpha version that is only used in Plone 6.2, you can specify that you don't want this version bump by adding the `--no-breaking` option.

```

I should probably add this logic to the `bumpversion` command from `zest.releaser`.

mauritsvanrees reviewed (commented):
Overall good, but I made some suggestions.

gforcada commented:
@mauritsvanrees thanks for all the tips and improvements! 🙇🏾 that's much better ✨ 

@stevepiercy I moved the file to the location you suggested and add a ToC entry.

Now is probably the time to improve the wording, which probably needs plenty 😅

stevepiercy commented:
@gforcada I requested another review from @mauritsvanrees, and I'd like @davisagli to review as well, before I make my final pass for non-technical stuff.

mauritsvanrees commented on docs/developer-guide/native-namespaces.md:
```diff
@@ -0,0 +1,201 @@
+
+As the switch to the native namespace has to be coordinated, all python distributions need to be only for the same Plone version, in this case it was decided to do it for the Plone 6.2 version.
+
+Thus, we need to ensure that the test matrix, only tests against this Plone version.
```
Superfluous comma.

```suggestion
Thus, we need to ensure that the test matrix only tests against this Plone version.
```

mauritsvanrees reviewed (dismissed):
LGTM.  One minor textual suggestion.

stevepiercy committed:
Update docs/developer-guide/native-namespaces.md

Co-authored-by: Maurits van Rees <maurits@vanrees.org>

davisagli reviewed (commented):
Content looks good to me

stevepiercy committed:
Add html_meta stuff

stevepiercy committed:
Condense sections and remove verbosity

stevepiercy committed:
Break sentence into two with relevant code blocks.
Break another code block into separate code blocks for ease of copy-pasting into a shell session without comments.

stevepiercy committed:
- Rename file to singular
- Break up single code block into multiple to make comments more legible and code copy-pasteable.
- Clean up grammar and enhance MyST syntax

stevepiercy reviewed (approved):
Thanks for the guide! I made a few minor enhancements for readability, grammar, and syntax. Will merge when green.

stevepiercy committed:
Merge branch '6.0' into 3928-native-namespace
